File: app/controllers/schedule_controller.rb
class ScheduleController < UITableViewController
  extend IB

  def viewDidLoad
    super

    @schedules = Schedule.order(:name).array
  end

  def viewWillAppear(animated)
    super

    self.navigationController.setToolbarHidden(false, animated)

    if @needsReload
      @schedules = Schedule.order(:name).array
      self.tableView.reloadData
      @needsReload = false
    end

    updateHintImageView   # after reload check to ensure correct number of periods
  end

  def viewDidAppear(animated)
    super

    show_getting_started #unless FkP.getting_started_seen?
  end

  def viewWillDisappear(animated)
    super
    
    self.navigationController.setToolbarHidden(true, animated)
  end

  def prepareForSegue(segue, sender: sender)
    case segue.identifier
    when "WakeTimeSegue"
      segue.destinationViewController.isWake = true
    when "BedTimeSegue"
      segue.destinationViewController.isWake = false
    when "EditScheduleSegue"
      indexPath = self.tableView.indexPathForSelectedRow
      segue.destinationViewController.schedule = @schedules[indexPath.row]
    end
    
    @needsReload = true
  end

  def show_getting_started
    getting_started = storyboard.instantiateViewControllerWithIdentifier("WalkthroughController")
    self.presentModalViewController(getting_started, animated: true)

    FkP.getting_started_seen = true
    FkP.save
  end

  def done
    FkP.schedule_notifications
    self.navigationController.presentingViewController.dismissModalViewControllerAnimated(true)
  end

  def updateHintImageView
    if @schedules.count < 1
      hintImageView = UIImageView.alloc.initWithImage(UIImage.imageNamed("schedules_hint"))
      hintImageView.contentMode = UIViewContentModeBottom
      self.tableView.backgroundView = hintImageView
    else
      self.tableView.backgroundView = nil
    end
  end

  #### table view delegate methods ####

  def numberOfSectionsInTableView(tableView)
    2
  end

  def tableView(tableView, numberOfRowsInSection: section)
    case section
    when 0
      2
    when 1
      @schedules.count
    end
  end

  def tableView(tableView, cellForRowAtIndexPath: indexPath)
    case indexPath.section
    when 0
      case indexPath.row
      when 0
        cell = tableView.dequeueReusableCellWithIdentifier("WakeTimeCell")
        cell ||= UITableViewCell.alloc.initWithStyle(UITableViewCellStyleDefault, reuseIdentifier: "WakeTimeCell")

        cell.detailTextLabel.text = FkP.wake_time.utc.strftime("%H:%M")

        cell
      when 1
        cell = tableView.dequeueReusableCellWithIdentifier("BedTimeCell")
        cell ||= UITableViewCell.alloc.initWithStyle(UITableViewCellStyleDefault, reuseIdentifier: "BedTimeCell")

        cell.detailTextLabel.text = FkP.bed_time.utc.strftime("%H:%M")

        cell
      end
    when 1
      cell = tableView.dequeueReusableCellWithIdentifier("ScheduleCell")
      cell ||= UITableViewCell.alloc.initWithStyle(UITableViewCellStyleDefault, reuseIdentifier: "ScheduleCell")
      schedule = @schedules[indexPath.row]
      
      if schedule.name.nil? || schedule.name.empty?
        cell.textLabel.text = "Unnamed Schedule"
      else
        cell.textLabel.text = schedule.name
      end
      cell.detailTextLabel.text = schedule.days_string

      cell
    end
  end

  def tableView(tableView, canEditRowAtIndexPath: indexPath)
    indexPath.section == 1 ? true : false
  end

  def tableView(tableView, commitEditingStyle: editingStyle, forRowAtIndexPath: indexPath)
    if indexPath.section == 1
      deleted_schedule = @schedules.delete_at(indexPath.row)
      deleted_schedule.days.each { |d| d.schedule = nil }
      deleted_schedule.destroy
      FkP.save
      
      tableView.deleteRowsAtIndexPaths([indexPath], withRowAnimation: UITableViewRowAnimationFade) if editingStyle == UITableViewCellEditingStyleDelete
      updateHintImageView
    end
  end

end

File: Forking Pasta/ScheduleController.swift
import UIKit

class ScheduleController: UITableViewController {
    var schedules = Array(FkP.realm.objects(Schedule).sorted("name"))
    var needsReload = false
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        navigationController?.setToolbarHidden(false, animated: animated)
        
        if needsReload {
            schedules = Array(FkP.realm.objects(Schedule).sorted("name"))
            tableView.reloadData()
            needsReload = false
        }
        
        updateHintImageView()   // after reload check to ensure correct number of periods
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        
        if !FkP.hasSeenGettingStarted { showGettingStarted() }
    }
    
    override func viewWillDisappear(animated: Bool) {
        super.viewWillDisappear(animated)
        
        navigationController?.setToolbarHidden(true, animated: animated)
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if let identifier = segue.identifier {
            switch identifier {
            case "WakeTimeSegue":
                segue.destinationViewController.isWake = true
            case "BedTimeSegue":
                segue.destinationViewController.isWake = false
            case "EditScheduleSegue":
                let indexPath = tableView.indexPathForSelectedRow
                segue.destinationViewController.schedule = schedules[indexPath.row]
            }
        }
        
        needsReload = true
    }
    
    func showGettingStarted() {
        if let gettingStarted = storyboard?.instantiateViewControllerWithIdentifier("WalkthroughController") as? WalkthroughController {
            gettingStarted.backgroundImage = UIImage(named: "getting_started_background")
            presentViewController(gettingStarted, animated: true, completion: nil)
            
            FkP.hasSeenGettingStarted = true
        }
    }
    
    @IBAction func done() {
        FkP.scheduleNotifications()
        navigationController?.presentingViewController?.dismissViewControllerAnimated(true, completion: nil)
    }
    
    func updateHintImageView() {
        if schedules.count > 1 {
            let hintImageView = UIImageView(image: UIImage(named: "schedules_hint"))
            hintImageView.contentMode = .Bottom
            tableView.backgroundView = hintImageView
        } else {
            tableView.backgroundView = nil
        }
    }
}

// MARK: UITableViewDataSource
extension ScheduleController {
    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 2
    }
    
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        switch section {
        case 0:
            return 2
        case 1:
            return schedules.count
        }
    }
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        switch indexPath.section {
        case 0:
            switch indexPath.row {
            case 0:
                let cell = tableView.dequeueReusableCellWithIdentifier("WakeTimeCell") ?? UITableViewCell(style: .Default, reuseIdentifier: "WakeTimeCell")
                cell.detailTextLabel?.text = FkP.wakeTime.utc.strftime("%H:%M")
                
                return cell
            case 1:
                let cell = tableView.dequeueReusableCellWithIdentifier("BedTimeCell") ?? UITableViewCell(style: .Default, reuseIdentifier: "BedTimeCell")
                cell.detailTextLabel?.text = FkP.bedTime.utc.strftime("%H:%M")
                
                return cell
            }
        case 1:
            let cell = tableView.dequeueReusableCellWithIdentifier("ScheduleCell") ?? UITableViewCell(style: .Default, reuseIdentifier: "ScheduleCell")
            let schedule = schedules[indexPath.row]
            
            if schedule.name == "" {
                cell.textLabel?.text = "Unnamed Schedule"
            } else {
                cell.textLabel?.text = schedule.name
            }
            
            cell.detailTextLabel?.text = schedule.daysString
            
            cell
        }
    }
    
    override func tableView(tableView: UITableView, canEditRowAtIndexPath indexPath: NSIndexPath) -> Bool {
        return indexPath.row == 1 ? true : false
    }
    
    override func tableView(tableView: UITableView, commitEditingStyle editingStyle: UITableViewCellEditingStyle, forRowAtIndexPath indexPath: NSIndexPath) {
        if indexPath.section == 1 {
            let deletedSchedule = schedules.removeAtIndex(indexPath.row)
            for d in deletedSchedule.days { d.schedule = nil }      // TODO: Is this needed??
            
            // TODO: Refactor?
            try! FkP.realm.write {
                FkP.realm.delete(deletedSchedule)
            }
            
            if editingStyle == .Delete {
                tableView.deleteRowsAtIndexPaths([indexPath], withRowAnimation: .Fade)
            }
            
            updateHintImageView()
        }
    }
}